StateDB in Cilium guide

joamaki opened the pull request:
An initial version of a guide on how/where/why StateDB is used in Cilium. The first commit fixes few metric related issues that came up while writing the docs.

ovidiutirla commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,379 @@
+~~~~~~~~~~~~
+
+`StateDB <https://github.com/cilium/statedb>`_ is an in-memory database developed
+for the Cilium project for management of control-plane state. It aims to simplify access
```
```suggestion
for the Cilium project to manage control-plane state. It aims to simplify access
```

ovidiutirla commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,379 @@
+lockless readers which improves resiliency compared to the RWMutex+hashmap+callback pattern 
+where a bug in a controller observing the state may cause critical functions to either stop or significantly
+decrease throughput. Additionally having flexible ways to access and index the state allows for opportunities to deduplicate
+state. Many components of the agent has historically functioned through callback-based subscriptions to
```
```suggestion
the state. Many components of the agent have historically functioned through callback-based subscriptions
```

ovidiutirla commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,379 @@
+The separation of state from logic operating on it (e.g. moving away from kitchen-sink `Manager` pattern)
+also opens up the ability to do wider and more meaningful integration testing on components of the agent. 
+When most of the inputs and outputs of a component are tables, we can combine multiple components into an
+integration test that is solely defined in terms of test inputs and expected outputs. This allows more validation to be performed with fairly simple integration tests rather than with the slower and costly end-to-end tests.
```
```suggestion
integration test that is solely defined in terms of test inputs and expected outputs. This allows more validation to be performed with fairly simple integration tests rather than with slower and costly end-to-end tests.
```

ovidiutirla commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,379 @@
+	:align: center
+
+
+The agent in this architectural style can be roughly thought of consisting of:
```
```suggestion
The agent in this architectural style can be broadly considered to consist of:
```

ovidiutirla commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,379 @@
+  contents of the desired state tables.
+
+- *Desired state tables*: the internal state that the controllers produce to succintly describe what should be done. For example a desired state table could describe what the contents of a BPF map should be or what routes should be installed.
+
```
```suggestion
- *Desired state tables*: the internal state that the controllers produce to succinctly describe what should be done. For example, a desired state table could describe what the contents of a BPF map should be or what routes should be installed.

```

ovidiutirla commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,379 @@
+
+* The control-plane of the agent is essentially everything outside the reconcilers This allows us to integration test, simulate or benchmark the control-plane code without unreasonable amount of scaffolding. The easier it is to write reliable integration tests the more resilient the codebase becomes.
+
+What we're trying to achieve is well summarized by Fred Brooks in "The Mythical Man Month":
+
+	| Show me your flowchart and conceal your tables, and I shall continue to be mystified.
+	| Show me your tables, and I won't usually need your flowchart; it'll be obvious.
```
❤️

ovidiutirla commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,379 @@
+	}
+
+	// TableName is a constant for the table name. This is used in cilium-dbg
+	// to refer to this table.
```
```suggestion
	// to refer to this table. The TableName must be composed of lowercase letters, separated by hyphens, and its total length cannot exceed X characters.
```
TODO, we should define a maximum limit of characters we can set on a table

joamaki replied:
please review: https://github.com/cilium/statedb/pull/52

ovidiutirla commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,379 @@
+				// Start tracking changes to the table. This instructs the database
+				// to keep deleted objects off to the side for us to observe.
+				wtxn := db.WriteTxn(table)
+				changes, err := table.Changes(wtxn)
```
Any particular reason we need the WriteTxn and ReadTxn is not enough? ( I know the interface signature requires WriteTxn but wondering why?)

txn usage in Changes()
* itxn := txn.getTxn() -> Taken from ReadTxn anyway
* t.Revision(txn) -> which taxes the ReadTxn
* t.LowerBoundWatch(txn..) -> same

joamaki replied:
The reason why `Changes` needs it is that it actually does modify the table:
https://github.com/cilium/statedb/blob/main/txn.go#L265. 

I did at some point think about it internally doing the WriteTxn, but in the end I think this way it's transparent, safer and it's also obvious to the user how this gets sequenced with other WriteTxn's. E.g. This locks the table and we'll see the objects deleted after we `Commit` this `WriteTxn`.

ovidiutirla reviewed (commented):
small nits, looks good so far.

joamaki committed:
hive: Fix StateDB and reconciler metrics

Relocating StateDB to its own repo reworked how metrics were done.
Unfortunately the new metrics implementation was not properly tested
as it was disabled by default.

This fixes the multiple issues it had:
- metrics.Metric cell was never defined for the metrics, so they were not
  registered.
- ConfigName dropped as it is automatically computed if missing. There were
  few that had wrong ConfigName.
- Fix use of "full" instead of "prune" in the pruning metrics (leftover from
  when full reconciliation split into refresh and prune).
- WithLabelValues was called with the label names instead of just the values
- Some labels were wrong way around
- Prune count wasn't incremented

The implementation was tested locally with "kind" by enabling all the metrics
with "prometheus.metrics" helm option and then checking "cilium metrics list".

Fixes: 4219e5c32926 ("treewide: Relocate pkg/statedb to cilium/statedb")
Signed-off-by: Jussi Maki <jussi@isovalent.com>

ovidiutirla commented on contrib/examples/statedb/example.go:
```diff
@@ -0,0 +1,134 @@
+	tick := time.NewTicker(time.Second)
+	defer tick.Stop()
+
+	health.OK("Starting")
```
would this be an interesting place to add a redirect to a page where we describe how `cell.Health` should be used?

joamaki replied:
Oh, looks like Guide to the Hive doesn't have a section for health. I'll add that and link from here.

joamaki replied:
I think I'll do a follow up on that instead of doing it in this PR to get this in as this already has fair bit of useful content.

ovidiutirla replied:
SGTM

ovidiutirla commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,348 @@
+TODO: update the dashboard from 
+https://github.com/joamaki/cilium/blob/pr/joamaki/bpf-ops-service-map/demo/
+
+StateDB
```
Should we keep all metrics in one place, ex. https://docs.cilium.io/en/stable/observability/metrics/?

joamaki replied:
I would probably hold off a bit from that as I expect these to still change and don’t want to give the wrong impression on their stability. But yes eventually they should go there (and be autogenerated)

joamaki committed:
Documentation: Guide for StateDB usage in Cilium

Add a guide for how StateDB is used in Cilium and what facilities
are available.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

ovidiutirla commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,443 @@
+
+* If the table is closely associated with a specific feature, define it alongside the implementation of the feature. If the table is shared by many modules, consider defining it in ``daemon/k8s`` or ``pkg/datapath/tables`` so it is easy to discover.
+
+* Make sure the object can be JSON marshalled so it can be inspected. If you need to store non-marshallable data (e.g. functions), make them private or mark them with ``json:"-"`` struct tag.
```
should private fields be printed to the CLI and have a placeholder or just be ignored?

joamaki replied:
Just ignored since they can't be transferred. If one needs to show information in a field that's not marshallable then you need a separate field to summarize that info for json marshalling (e.g. for `cilium-dbg statedb` commands and dumping).

ovidiutirla commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,443 @@
+    Pod(kube-system/cilium-envoy-8xwp7): Running (revision: 2, deleted: false)
+    ...
+
+Reconcilers
```
Should we add some details around the difference of using table pointers in context of reconcilers?

joamaki replied:
Yes! I'll add some points about it.

ovidiutirla commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,443 @@
+
+To set up the reconciler you will need the following.
+
+Add ``reconciler.Status`` as a field into your object (there can be multiple):
```
what about the index for the status?

joamaki replied:
Not needed usually, unless the upside of quickly looking up e.g. pending objects outweigh downsides/overhead of maintaining a separate status index. There has not been a case like that yet so I wouldn't document it.

nebril reviewed (approved):
ci-structure LGTM!

qmonnet commented on Documentation/spelling_wordlist.txt:
```diff
@@ -103,6 +103,8 @@ Polytechnique
 QCon
 QLogic
+Reconciler
+Reconcilers
```
These aren't necessary given that you also added the lower-case (and thus, case-insensitive) version for these words.

joamaki replied:
dropped.

qmonnet commented on Documentation/contributing/development/_static/statedb-arch.svg:
There seems to be a very high bottom margin to the file, below the actual picture. Is this intentional?

joamaki replied:
I don't see that here. Can you paste a screenshot?

qmonnet replied:
![image](https://github.com/user-attachments/assets/1f72e280-4042-45e9-b5cd-ffe9922b57fc)

Could be just GitHub's preview engine, though. I didn't check the file locally, or checked the preview for the docs.

joamaki replied:
I think it's that. It renders correctly in the docs.

qmonnet commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,443 @@
+We assume familiarity with the Hive framework. If you're not familiar with it,
+consider reading through :ref:`guide-to-the-hive` first.
+
+The Why
```
```suggestion
Motivation
```

joamaki replied:
👍

qmonnet commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,443 @@
+
+You can find and run the above examples in ``contrib/examples/statedb``:
+
+.. code-block:: shell
```
```suggestion
.. code-block:: shell-session
```

qmonnet commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,443 @@
+
+We can now dump the contents of our example table:
+
+.. code-block:: shell
```
```suggestion
.. code-block:: shell-session
```

Although the pound sign will mess up formatting a bit (it will be considered as a prompt symbol).

qmonnet commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,443 @@
+.. code-block:: shell
+
+    contrib/examples/statedb_k8s$ go run . --k8s-kubeconfig-path ~/.kube/config
+    level=info msg=Starting
```
```suggestion
.. code-block:: shell-session
    $ contrib/examples/statedb_k8s$ go run . --k8s-kubeconfig-path ~/.kube/config
    level=info msg=Starting
```

qmonnet replied:
Woops sorry I didn't mean to remove the blank line separator :grimacing:

qmonnet replied:
Looks like you kept it though :relieved:

qmonnet commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,443 @@
+  myObjects.Insert(wtxn, &MyObject{ID: 123, Status: reconciler.StatusPending()})
+  wtxn.Commit()
+
+The reconciler watches the tables (using ``Changes()``) and will call ``Update`` for each
```
```suggestion
The reconciler watches the tables (using ``Changes()``) and calls ``Update`` for each
```

Nit: We recommend using present tense

qmonnet commented on Documentation/contributing/development/statedb.rst:
```diff
@@ -0,0 +1,443 @@
+    - +cilium_reconciler_prune_duration_seconds
+
+These are still under development and the metric names may change. We also do not
+as of yet have a grafana dashboard template for these.
```
“as of yet” is tricky in docs. How about creating an issue to track this and pointing to it in the docs?

joamaki replied:
I'll drop this. It's not yet entirely clear if/how/when we'll do a dashboard for the reconciler metrics.

qmonnet reviewed (changes requested):
Looks good, thanks! With just a few small nits from my side.

pippolo84 commented on Documentation/contributing/development/_static/statedb-arch.d2:
```diff
@@ -0,0 +1,72 @@
+# Regenerate with:
+#   (outside cilium/cilium workspace)
+#   $ go install oss.terrastruct.com/d2@latest
```
because of a replace directive in its go.mod, `d2` latest version won't be installed with `go install`:

```
$ go install oss.terrastruct.com/d2@latest
go: oss.terrastruct.com/d2@latest (in oss.terrastruct.com/d2@v0.6.7):
	The go.mod file for the module providing named packages contains one or
	more replace directives. It must not contain directives that would cause
	it to be interpreted differently than if it were the main module.
```

I guess it will be fixed in next releases, but wondering if we should just give more generic instructions like "install `d2` following instructions at https://github.com/terrastruct/d2/"

joamaki replied:
sgtm, reworded.

pippolo84 reviewed (approved):
LGTM

qmonnet reviewed (approved):
Thanks!
